### PR TITLE
Build Clipper with `tools=no` and patch it to auto-disable exceptions

### LINKED ISSUE #25262: Compiling with Javascript tools=yes fails (no exceptions support for Clipper and Assimp build)
**Godot version:**
master (https://github.com/godotengine/godot/commit/2c9536b655d006c697a3cf830ad3352f1a1ed9f2)

**OS/device including version:**
Arch Linux

**Issue description:**
The javascript export (html5) fails with tools=yes.
I think it should work because there was the website http://godot.online.
```
[ 72%] Compiling ==> editor/collada/collada.cpp
thirdparty/misc/clipper.cpp:901:7: error: cannot use 'throw' with exceptions
      disabled
      throw clipperException("Coordinate outside allowed range");
      ^
thirdparty/misc/clipper.cpp:1049:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("AddPath: Open paths must be subject.");
    ^
thirdparty/misc/clipper.cpp:1081:5: error: cannot use 'throw' with exceptions
      disabled
    throw; //range test fails
    ^
thirdparty/misc/clipper.cpp:1065:3: error: cannot use 'try' with exceptions
      disabled
  try
  ^
thirdparty/misc/clipper.cpp:1445:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("UpdateEdgeIntoAEL: invalid call");
    ^
thirdparty/misc/clipper.cpp:1513:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("Error: PolyTree struct is needed for open pa...
    ^
thirdparty/misc/clipper.cpp:1563:3: error: cannot use 'try' with exceptions
      disabled
  try {
  ^
thirdparty/misc/clipper.cpp:2841:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("ProcessIntersections error");
    ^
thirdparty/misc/clipper.cpp:2830:3: error: cannot use 'try' with exceptions
      disabled
  try {
  ^
thirdparty/misc/clipper.cpp:3005:8: error: cannot use 'throw' with exceptions
      disabled
  else throw clipperException("DoMaxima error");
       ^
10 errors generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
scons: *** [thirdparty/misc/clipper.javascript.opt.tools.64.bc] Error 1
scons: building terminated because of errors.
```

**Steps to reproduce:**
```scons p=javascript tools=yes target=release_debug bits=64 -j 8```

### PR DESCRIPTION
**Pull request by Xrayez**
Prerequisite for #28987.
Reverts #17319.
~Helps #25262 (needs testing) seems to be fixed for Clipper but other thirdparty code with similar problem detected (assimp), which also uses exceptions.~ already fixed by #29032.

### CODE REVIEW

**Comment by Xrayez** [comment]
Ok, got this tested and works with sprite editor plugin as before, works alright with #28987 in both `release_debug` and `release` targets under Windows. I can squash commits together but as it's touching thirdparty code, waiting for review. "Patch" is generated with `git diff` against working tree.

File: editor/SCsub
Comment (akien-mga) [inline editor/SCsub]: The whole `env_thirdparty` is unused now.

File: thirdparty/misc/clipper.cpp
Comment (akien-mga) [inline thirdparty/misc/clipper.cpp]: Encoding change?

File: thirdparty/misc/clipper.hpp
Comment (akien-mga) [inline thirdparty/misc/clipper.hpp]: I guess you can also remove `#include <stdexcept>`?

**Comment by akien-mga** [comment]
I'd prefer not to have to modify upstream code at all, unless we really need to. In this case, this should ideally be implemented upstream in a way that lets us disable exceptions, similar to https://github.com/nlohmann/json/blob/ee4028b8e4cf6a85002a296a461534e2380f0d57/single_include/nlohmann/json.hpp#L518-L530

But as there's little hope of seeing any new Clipper 6.x release, I guess we don't have much choice. Commits should be squashed, either all together or keeping the revert separate; but given that this revert by itself reintroduces a bug, I guess it's better to include it together with the rest of the changes which prevent the bug from happening.

File: editor/SCsub
Comment (Xrayez) [reply editor/SCsub]: Hmm, well then it's strange why it compiles without linking errors.

File: thirdparty/misc/clipper.cpp
Comment (Xrayez) [reply thirdparty/misc/clipper.cpp]: This is weird one because `git diff` makes encoding look like this, but the encoding could still be preserved while committing changes... so I left it this way.

Setting `LESSCHARSET=UTF-8 git diff` doesn't seem to cure this either.

**Comment by Xrayez** [comment]
Angus Johnson is currently working on Clipper 10.0.0. Clipper was written in Delphi originally and translated to C++ pretty much mechanically as the author admits that his C++ knowledge is somewhat limited, so it could take quite some time to propagate changes from upstream for such C++ specific change that he has little knowledge of. I've participated in speeding up Clipper 10.0.0 C++ translation as well so he could delegate this work to me (which would be easier for me just to patch this manually? 😄).

**Comment by Xrayez** [comment]
Not related to 6.4.2 but I've tried to translate these exception fixes ideas already to Angus Johnson [here](https://sourceforge.net/p/polyclipping/bugs/192/#e916/cd04/5e97/3bd0/6b4b/65c4/eb0b/4004/f30a/e869/faae).

File: editor/SCsub
Comment (akien-mga) [reply editor/SCsub]: Well I mean it does nothing, you remove the only file that was built within `env_thirdparty` which is created two lines above. So you should remove it.

File: thirdparty/misc/clipper.cpp
Comment (akien-mga) [reply thirdparty/misc/clipper.cpp]: `git add -p` should let you selectively add changes and thus ignore those if they're generated by your OS and you can't find how to fix them (I've had similar issues in the past too with weird Mac line endings).

Comment (Xrayez) [reply thirdparty/misc/clipper.cpp]: Used `git add -p`, `VSCode` seems to mess up encoding. 🐛

Comment (akien-mga) [reply thirdparty/misc/clipper.cpp]: Talking about tools bug, it's weird that your `.patch` file appears as a binary blob in the GitHub diff. In my experience GitHub used to recognize those as text files, but maybe only when created on Unix? See e.g. https://github.com/godotengine/godot/pull/27929/commits/668439d16a5c30587ab77461fb3211306ad62587

Comment (Xrayez) [reply thirdparty/misc/clipper.cpp]: I checked the encoding for those, and the patch in 668439d appears as UTF-8 LF, while the patch generated in Windows Git generates UTF-16LE CRLF, but maybe because it's created locally and wasn't checked out from remote yet.

Comment (akien-mga) [reply thirdparty/misc/clipper.cpp]: #29098 should fix the encoding issue with these comments, you might have to fight git a bit when rebasing, but hopefully it should be fine since your PR no longer changes this code chunk.

I also pushed a change to enforce LF line endings in the git repo (converted automatically to CRLF when checking out on Windows, and converted back to LF when committing) for all text files, including `.patch` files, so this might solve the diff issue.

Comment (Xrayez) [reply thirdparty/misc/clipper.cpp]: Yeah there's plenty of changes so far, I'm going to rebase once it's merged.

**Comment by Xrayez** [comment]
I proposed preliminary changes to upstream here with the patch anyway: 
https://sourceforge.net/p/polyclipping/bugs/193/

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
